presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java

```diff
@@ -76,42 +88,44 @@ public Parts createPartialAggregations(PlanVariableAllocator variableAllocator,
                     originalAggregation.getOrderBy(),
                     originalAggregation.isDistinct(),
                     originalAggregation.getMask()));
-            finalAggregation.put(entry.getKey(),
+
+            // create final aggregation
+            finalOrIntermediateAggregations.put(entry.getKey(),
                     new Aggregation(
                             new CallExpression(
                                     originalAggregation.getCall().getDisplayName(),
                                     functionHandle,
-                                    function.getFinalType(),
+                                    intermediate ? function.getIntermediateType() : function.getFinalType(),
                                     ImmutableList.of(partialVariable)),
                             Optional.empty(),
                             Optional.empty(),
                             false,
                             Optional.empty()));
         }
         return new Parts(
-                new StatisticAggregations(partialAggregation.build(), groupingVariables),
-                new StatisticAggregations(finalAggregation.build(), groupingVariables));
+                new StatisticAggregations(finalOrIntermediateAggregations.build(), groupingVariables),
+                new StatisticAggregations(partialAggregations.build(), groupingVariables));
     }
 
     public static class Parts
     {
+        private final StatisticAggregations finalOrIntermediateAggregation;
```

Comment:
hmm.. my understanding is:

- partial aggr: (state, input) -> state
- intermediate aggr: (state, state) -> state
- final aggr: (state) -> output

So have final or intermediate aggregation store in same variable sounds a bit strange to me. Probably something I didn't understand correctly?

Reply:
> hmm.. my understanding is:

Yes, that is correct.

> So have final or intermediate aggregation store in same variable sounds a bit strange to me

I just didn't want to duplicate code for splitting an aggregation into `PARTIAL` + `FINAL`  and `PARTIAL` + `INTERMIDATE`.  I'm not sure if calling these fields `first` and `second` or `upstream` and `downstream` will be any better.